lib/AST/LayoutConstraint.cpp

```diff
@@ -114,4 +143,74 @@ void *LayoutConstraintInfo::operator new(size_t bytes, const ASTContext &ctx,
 
 SourceRange LayoutConstraintLoc::getSourceRange() const { return getLoc(); }
 
+#define MERGE_TO(lhs, rhs, result)                                             \
+  mergeTable[int(LayoutConstraintKind::lhs)][int(LayoutConstraintKind::rhs)] = \
+      LayoutConstraintKind::result
+
+#define MERGE_LOOKUP(lhs, rhs)                                                 \
+  mergeTable[int(lhs)][int(rhs)]
+
+LayoutConstraint LayoutConstraint::merge(LayoutConstraint Other) {
+  /// Define a mapping
+  // (LayoutConstraintKind, LayoutConstraintKind) -> LayoutConstraintKind
+  // It shows what kind of layout constraint would be produced by merging of two
+  // layout constraints with given kinds.
+  static LayoutConstraintKind
+      mergeTable[unsigned(LayoutConstraintKind::LastLayout) + 1]
+                [unsigned(LayoutConstraintKind::LastLayout) + 1];
+  static int initializedMergeTable;
+
+  if (!initializedMergeTable) {
+    // initialize the merge table once.
+    initializedMergeTable = true;
+    MERGE_TO(TrivialOfAtMostSize, Trivial, TrivialOfAtMostSize);
```

Comment:
You should be able to use initializer list nesting `{ { ... }, { ... }, ... }` to avoid the imperative initialization and `initializedMergeTable` bit. It'll also make sure that cases never get skipped, e.g., the missing `merge(TrivialOfExactSize, TrivialOfExactSize)` case.